mcp/tools/tools_coder.go

```diff
@@ -0,0 +1,381 @@
+// {"jsonrpc":"2.0","id":1,"method":"tools/call", "params": {"name": "coder_workspace_exec", "arguments": {"workspace": "dev", "command": "ps -ef"}}}
+func handleCoderWorkspaceExec(deps ToolDeps) mcpserver.ToolHandlerFunc {
+	return func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
+		if deps.Client == nil {
+			return nil, xerrors.New("developer error: client is required")
+		}
+		args := request.Params.Arguments
+
+		wsArg, ok := args["workspace"].(string)
+		if !ok {
+			return nil, xerrors.New("workspace is required")
+		}
+
+		command, ok := args["command"].(string)
+		if !ok {
+			return nil, xerrors.New("command is required")
+		}
+
+		// Validate the command if allowed commands are specified
+		allowed, err := IsCommandAllowed(command, deps.AllowedExecCommands)
+		if err != nil {
+			return nil, err
+		}
+		if !allowed {
+			return nil, xerrors.Errorf("command not allowed: %s", command)
+		}
+
+		// Attempt to fetch the workspace. We may get a UUID or a name, so try to
+		// handle both.
+		ws, err := getWorkspaceByIDOrOwnerName(ctx, deps.Client, wsArg)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to fetch workspace: %w", err)
+		}
+
+		// Ensure the workspace is started.
+		// Select the first agent of the workspace.
+		var agt *codersdk.WorkspaceAgent
+		for _, r := range ws.LatestBuild.Resources {
+			for _, a := range r.Agents {
+				if a.Status != codersdk.WorkspaceAgentConnected {
+					continue
+				}
+				agt = ptr.Ref(a)
+				break
+			}
+		}
+		if agt == nil {
+			return nil, xerrors.Errorf("no connected agents for workspace %s", ws.ID)
+		}
+
+		conn, err := workspacesdk.New(deps.Client).AgentReconnectingPTY(ctx, workspacesdk.WorkspaceAgentReconnectingPTYOpts{
+			AgentID:     agt.ID,
+			Reconnect:   uuid.New(),
+			Width:       80,
+			Height:      24,
+			Command:     command,
+			BackendType: "buffered", // the screen backend is annoying to use here.
+		})
+		if err != nil {
+			return nil, xerrors.Errorf("failed to open reconnecting PTY: %w", err)
+		}
+		defer conn.Close()
+
+		var buf bytes.Buffer
+		if _, err := io.Copy(&buf, conn); err != nil {
+			// EOF is expected when the connection is closed.
+			// We can ignore this error.
+			if !errors.Is(err, io.EOF) {
+				return nil, xerrors.Errorf("failed to read from reconnecting PTY: %w", err)
+			}
+		}
+
+		return &mcp.CallToolResult{
+			Content: []mcp.Content{
+				mcp.NewTextContent(strings.TrimSpace(buf.String())),
+			},
+		}, nil
+	}
+}
+
+// Example payload:
+// {"jsonrpc":"2.0","id":1,"method":"tools/call", "params": {"name": "coder_list_templates", "arguments": {}}}
+func handleCoderListTemplates(deps ToolDeps) mcpserver.ToolHandlerFunc {
+	return func(ctx context.Context, _ mcp.CallToolRequest) (*mcp.CallToolResult, error) {
+		if deps.Client == nil {
+			return nil, xerrors.New("developer error: client is required")
+		}
+		templates, err := deps.Client.Templates(ctx, codersdk.TemplateFilter{})
```

Comment:
is there risk of pagination?

Reply:
Right now, that endpoint is not paginated. There are some query parameters we could pass in, such as `deleted`, `organization_id`, `exact_name` etc.